### PR TITLE
feat: add observe status access-key for pre-check and logging only (#5216)

### LINKED ISSUE #5216: 增加未启用密钥的预校验能力，方便启用密钥前观察客户端的行为是否服务预期
**你的特性请求和某个问题有关吗？请描述**

1. 背景：使用 apollo 前期没有开启密钥功能，现在想要开启密钥功能，目前的规则是 “一旦有启用的访问密钥，**客户端将被要求配置密钥，否则无法获取配置**”
2. 遇到的问题：客户端经过改造完成后，正式启用密钥前，没有**简单直接的方式判断所有的客户端的行为是否符合预期**
4. 特性诉求：因此希望 apollo 提供一种对未启用密钥的预校验能力，不阻塞请求，但是对于未正确携带密钥的客户端访问进行日志记录，方便做正式启用密钥前的 check

**清晰简洁地描述一下你希望的解决方案**

1. 增加开关配置标识是否开启密钥的预校验能力，默认不开启（`兼容性`） 
2. 密钥的预校验能力大概逻辑如下
    1. 触发条件：`开关开启` + `有密钥配置` + `所有密钥均未开启`
    2. 触发后行为：打印日志，日志至少包含 `appId`, `clientIp`, `authorization` 等信息 

**BTW**：如果认为这是一个合理的诉求，这边可以提一个 PR 进行实现

### PR DESCRIPTION
**Pull request by larry4xie**
## What's the purpose of this PR

add observe status access-key for pre-check and logging only

## Which issue(s) this PR fixes:
Fixes #5216

## Brief changelog

image

image

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Read the [Contributing Guide](https://github.com/apolloconfig/apollo/blob/master/CONTRIBUTING.md) before making this pull request.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit tests to verify the code.
- [x] Run `mvn clean test` to make sure this pull request doesn't break anything.
- [x] Update the [`CHANGES` log](https://github.com/apolloconfig/apollo/blob/master/CHANGES.md).

## Release Notes

- **New Features**
	- Introduced a new `mode` parameter for enabling and disabling access keys, allowing users to specify operational modes (filter or observer).
	- Added a method to retrieve observable secrets based on application ID, expanding access key functionalities.
	- Enhanced authentication checks with improved logging for better tracking and security.
	- Updated release notes to reflect new features and enhancements for version 2.4.0.

- **Bug Fixes**
	- Improved logging and validation mechanisms in client authentication for better tracking and security.

- **Database Changes**
	- Added a new `Mode` column to the `AccessKey` table, specifying key modes for enhanced access key management.

These changes enhance user control and clarity in managing access keys within the application.

### CODE REVIEW

File: apollo-adminservice/src/main/java/com/ctrip/framework/apollo/adminservice/controller/AccessKeyController.java [hunk 1]
```diff
@@ -61,9 +63,11 @@ public void delete(@PathVariable String appId, @PathVariable long id, String ope
   }
 
   @PutMapping(value = "/apps/{appId}/accesskeys/{id}/enable")
-  public void enable(@PathVariable String appId, @PathVariable long id, String operator) {
+  public void enable(@PathVariable String appId, @PathVariable long id,
+      @RequestParam(required = false, defaultValue = "0") int mode, String operator) {
```
Comment (nobodyiam) [inline apollo-adminservice/src/main/java/com/ctrip/framework/apollo/adminservice/controller/AccessKeyController.java:67]: Is this clearer?
```suggestion
      @RequestParam(required = false, defaultValue = ""+AccessKeyMode.FILTER) int mode, String operator) {
```

File: apollo-configservice/src/main/java/com/ctrip/framework/apollo/configservice/filter/ClientAuthenticationFilter.java [hunk 2]
```diff
@@ -68,26 +70,49 @@ public void doFilter(ServletRequest req, ServletResponse resp, FilterChain chain
...
+      if (!checkAuthorization(authorization, availableSecrets, timestamp,
+          request.getRequestURI(), request.getQueryString())) {
+        logger.warn("Invalid authorization. appId={},clientIp={},authorization={}",
+            appId, ip, authorization);
         response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Unauthorized");
         return;
       }
+    } else {
+      // pre-check timestamp and signature
+      List<String> observableSecrets = accessKeyUtil.findObservableSecrets(appId);
```
Comment (nobodyiam) [inline apollo-configservice/src/main/java/com/ctrip/framework/apollo/configservice/filter/ClientAuthenticationFilter.java:97]: The observable check logic closely resembles the filter logic. Consider extracting it into a method that takes a mode variable.

File: apollo-configservice/src/main/java/com/ctrip/framework/apollo/configservice/filter/ClientAuthenticationFilter.java [hunk 3]
```diff
@@ -130,4 +155,11 @@ private boolean checkAuthorization(String authorization, List<String> availableS
     }
     return false;
   }
+
+  private void preCheckInvalidLogging(String message) {
+    logger.warn(message);
+    Tracer.logEvent("Apollo.AccessKey.PreCheck", message);
+    // only for test mock
+    accessKeyUtil.preCheckInvalid();
```
Comment (nobodyiam) [inline apollo-configservice/src/main/java/com/ctrip/framework/apollo/configservice/filter/ClientAuthenticationFilter.java:163]: We could change the method to `protected` and use `Spy` on the filter in the unit test to verify.

```java
clientAuthenticationFilter = spy(new ClientAuthenticationFilter(bizConfig, accessKeyUtil));
...
verify(clientAuthenticationFilter, never()).preCheckInvalidLogging(anyString());
```

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/controller/AccessKeyController.java [hunk 4]
```diff
@@ -77,9 +77,10 @@ public void delete(@PathVariable String appId,
   @ApolloAuditLog(type = OpType.UPDATE, name = "AccessKey.enable")
   public void enable(@PathVariable String appId,
       @PathVariable String env,
-      @PathVariable long id) {
+      @PathVariable long id,
+      @RequestParam(required = false, defaultValue = "0") int mode) {
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/controller/AccessKeyController.java:81]:
```suggestion
      @RequestParam(required = false, defaultValue = ""+AccessKeyMode.FILTER) int mode) {
```

**Review by nobodyiam — COMMENTED** [review]
This looks great! Please find some comments below.

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/controller/AccessKeyController.java [hunk 5]
```diff
@@ -77,9 +79,10 @@ public void delete(@PathVariable String appId,
   @ApolloAuditLog(type = OpType.UPDATE, name = "AccessKey.enable")
   public void enable(@PathVariable String appId,
       @PathVariable String env,
-      @PathVariable long id) {
+      @PathVariable long id,
+      @RequestParam(required = false, defaultValue = "" + FILTER) int mode) {
```
[the remark this answers is not in the document]
Comment (larry4xie) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/controller/AccessKeyController.java:83]: Attribute value must be constant

**Comment by larry4xie** [comment]
> This looks great! Please find some comments below.

👍🏻 All the comments have been resolved, Please take a moment to review again.

**Review by nobodyiam — COMMENTED** [review]
Please also update the CHANGES.md.

**Comment by larry4xie** [comment]
> Please also update the CHANGES.md.

Done.

**Review by nobodyiam — APPROVED** [review]
LGTM

**Pull request merged**